## Background

In the fastai course Part 1 [Lesson 6 video](https://youtu.be/AdhG64NF76E?feature=shared&t=3117) Jeremy Howard walked through the notebooks [First Steps: Road to the Top, Part 1](https://www.kaggle.com/code/jhoward/first-steps-road-to-the-top-part-1) and [Small models: Road to the Top, Part 2](https://www.kaggle.com/code/jhoward/small-models-road-to-the-top-part-2) where he builds increasingly accurate solutions to the [Paddy Doctor: Paddy Disease Classification](https://www.kaggle.com/competitions/paddy-disease-classification) Kaggle Competition. In the video, Jeremy referenced a series of walkthrough videos that he made while working through the four-notebook series for this competition. I'm excited to watch these walkthroughs to better understand how to approach a Kaggle competition from the perspective of a former #1 Kaggle grandmaster.

In this blog post series, I'll walk through the code Jeremy shared in each of the 6 Live Coding videos focused on this competition, submitting predictions to Kaggle along the way. My last two blog posts in this series reference Jeremy's [Scaling Up: Road to the Top, Part 3](https://www.kaggle.com/code/jhoward/scaling-up-road-to-the-top-part-3) notebook to improve my large model ensemble predictions. Here are the links to each of the blog posts in this series:

- [Part 1: Live Coding 8](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-1/)
- [Part 2: Live Coding 9](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-2/)
- Part 3: Live Coding 10 (You are here)
- [Part 4: Live Coding 11](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-4/)
- [Part 5: Live Coding 12](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-5/)
- [Part 6: Live Coding 13](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-6/)
- [Part 7: Improving My Large Ensemble, Part 1](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-7/)
- [Part 8: Improving My Large Ensemble, Part 2](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-8/)


[Link to the Live Coding 10 video](https://www.youtube.com/watch?v=zhBRynq9Yvo)

## Setup

In [5]:
!pip install -qq timm==0.6.13
import timm
timm.__version__

'0.6.13'

In [6]:
# install fastkaggle if not available
try: import fastkaggle
except ModuleNotFoundError:
    !pip install -Uq fastkaggle

from fastkaggle import *
from fastai.vision.all import *

comp = 'paddy-disease-classification'
path = setup_comp(comp, install='fastai')

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [7]:
path.ls()

(#4) [Path('../input/paddy-disease-classification/sample_submission.csv'),Path('../input/paddy-disease-classification/train_images'),Path('../input/paddy-disease-classification/train.csv'),Path('../input/paddy-disease-classification/test_images')]

In [8]:
trn_path = path/'train_images'

## The best models for fine tuning image recognition

All deep learning models will return a set of probabilities. That's what their final layer returns and we decode them using `argmax` across them. There's nothing to stop you from using those probabilities directly.

The Paddy Classification competition are kind of like the natural images you see in ImageNet, but ImageNet doesn't have any categories about diseases, they have categories about what's the main object in this image, such as different types of grass or fields or something. It's a bit different to ImageNet, which is what most of our pretrained models are trained on. Nearly all of the images are the same shape and size in this competition. 

There are two key dimensions that really seem to impact how well a model can be fine-tuned:
- How similar is your dataset to the dataset used for the pretrained model? 
  - If it's similar (like PETS to ImageNet), then the critical factor is how well does the fine-tuning of the model maintain the weights that are pretrained. They're probably not going to change very much. And you can take advantage of really big accurate models that have learned to do almost the exact same thing that you are trying to do.
  - If it's not similar (like Planets to ImageNet), a lot of the weights of a pretrained model are going to be useless for fine-tuning this because they'e learned specific features (like what does text look like, what do eyeballs look like, what does fur look like) none of which are going to be useful at all. 
- How big is your dataset?
  - On a big dataset, you've got time and epochs to take advantage of having lots of parameters in the model to learn to use them effectively. If you don't have much data you don't have much ability to do that.
  
  
Jeremy and Thomas Capelle analyze which models are the best for fine-tuning and Jeremy published the results in [this notebook](https://www.kaggle.com/code/jhoward/the-best-vision-models-for-fine-tuning). They used [YAML files](https://github.com/tcapelle/fastai_timm/blob/main/sweep_planets_lr.yaml) for Weights and Biases to define the different models and parameters that they wanted to test. You can use the wandb web GUI to view the training results. This [gist](https://gist.githubusercontent.com/jph00/959aaf8695e723246b5e21f3cd5deb02/raw/sweep.csv) has the results.

You can export a `pandas.DataFrame` to a `StringIO()` object which essentially stores the data as a string.

```python
from io import StringIO
strm = StringIO()
df.to_csv(strm, index=False)
txt = strm.getvalue()
```

You can also create a gist programatically:

```python
import ghapi.core as gh
g = gh.GhApi()
gist = g.create_gist('name', txt, filename='name.csv', public=True)

# view URL
gist.html_url
```

The `vit` family of models is particularly good at rapidly identifying features of data types it hasn't seen before (like medical imaging or satellite imagery). They also have good error rate with low memory usage. The `swin` family, also a transformers-based model like `vit`, was the most accurate for fine-tuning the Planets dataset. For the Planets dataset, the really big slow models don't necessarily have better error rates. Which makes sense because if they have heaps of parameters but they're trying to learn something they've never seen before it's unlikely that we will be able to take advantage of those parameters.

For some models (like vit_small_patch16_224) you can only use 224x224 image sizes, while with others (like convnext_tiny) you can use any sized images.

Jeremy ran the vision model fine-tuning on 3 RTX GPUs for about 12 hours. They didn't try all combinations of all parameters. Thomas ran a learning rate sweep to get a sense of what learning rates work well, and then they tried a couple of learning rates, a couple of the best resize methods and a couple of the best pooling types across a few broadly different kinds of models across the two different datasets. In every single case, the same learning rate, resize method and pooling method was the best. 

## Applying learning on Paddy notebook with small models

Let's try out some of these models for the paddy classification task to identify which ones' larger versions we should try training next. We use a fixed validation seed (`seed=42`) so that the same validation set is created each time we run `train`. The final batch size in a convnext model is 32x32 you generally you want both sides of the image to be sized in multiples of 32. The correct dimensions for `Resize` is `640` by `480`.

In [5]:
def train(arch, item, batch, accum=False):
    kwargs = {'bs': 16} if accum else {}
    dls = ImageDataLoaders.from_folder(trn_path, seed=42, valid_pct=0.2, item_tfms=item, batch_tfms=batch, **kwargs)
    cbs = GradientAccumulation(4) if accum else []
    learn = vision_learner(dls, arch, metrics=error_rate, cbs=cbs).to_fp16()
    learn.fine_tune(12, 0.01)
    return error_rate(*learn.tta(dl=dls.valid))

In [6]:
arch = 'convnext_small_in22k'

In [8]:
train(arch, item=Resize(480, method='squish'), batch=aug_transforms(size=224, min_scale=0.75))

Downloading: "https://dl.fbaipublicfiles.com/convnext/convnext_small_22k_224.pth" to /root/.cache/torch/hub/checkpoints/convnext_small_22k_224.pth


TensorBase(0.0211)

In [9]:
train(arch, item=Resize(480), batch=aug_transforms(size=224, min_scale=0.75))

TensorBase(0.0202)

In [10]:
train(arch, item=Resize((640,480)), batch=aug_transforms(size=(288,224), min_scale=0.75))

TensorBase(0.0187)

In [11]:
train(arch, item=Resize((640,480)), batch=aug_transforms(size=(320,240), min_scale=0.75))

TensorBase(0.0226)

In [12]:
arch = 'vit_small_patch16_224'

In [13]:
train(arch, item=Resize(480, method='squish'), batch=aug_transforms(size=224, min_scale=0.75))

TensorBase(0.0245)

In [14]:
train(arch, item=Resize(480), batch=aug_transforms(size=224, min_scale=0.75))

TensorBase(0.0250)

In [15]:
train(arch, item=Resize(640, method=ResizeMethod.Pad, pad_mode=PadMode.Zeros), batch=aug_transforms(size=224, min_scale=0.75))

TensorBase(0.0221)

In [17]:
arch = 'swinv2_base_window12_192_22k'

In [18]:
train(arch, item=Resize(480, method='squish'), batch=aug_transforms(size=192, min_scale=0.75))

/opt/conda/lib/python3.10/site-packages/torch/functional.py:504: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at /usr/local/src/pytorch/aten/src/ATen/native/TensorShape.cpp:3483.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Downloading: "https://github.com/SwinTransformer/storage/releases/download/v2.0.0/swinv2_base_patch4_window12_192_22k.pth" to /root/.cache/torch/hub/checkpoints/swinv2_base_patch4_window12_192_22k.pth


TensorBase(0.0173)

In [19]:
train(arch, item=Resize(480), batch=aug_transforms(size=192, min_scale=0.75), accum=True)

TensorBase(0.0183)

In [20]:
train(arch, item=Resize(640, method=ResizeMethod.Pad, pad_mode=PadMode.Zeros), batch=aug_transforms(size=192, min_scale=0.75), accum=True)

TensorBase(0.0192)

In [23]:
arch = 'swin_small_patch4_window7_224'

In [24]:
train(arch, item=Resize(480, method='squish'), batch=aug_transforms(size=224, min_scale=0.75))

Downloading: "https://github.com/SwinTransformer/storage/releases/download/v1.0.0/swin_small_patch4_window7_224.pth" to /root/.cache/torch/hub/checkpoints/swin_small_patch4_window7_224.pth


TensorBase(0.0207)

In [25]:
train(arch, item=Resize(480), batch=aug_transforms(size=224, min_scale=0.75))

TensorBase(0.0231)

In [26]:
train(arch, item=Resize(640, method=ResizeMethod.Pad, pad_mode=PadMode.Zeros), batch=aug_transforms(size=224, min_scale=0.75))

TensorBase(0.0183)

I'll summarize the training run parameters and resulting TTA error rates on the validation set in the following table. I have sorted this table by model name and descening TTA Error Rate (First Run).

|Architecture|item_tfms|batch_tfms|TTA Error Rate (First Run)|Minutes per epoch (First Run)|TTA Error Rate (Second Run)|
|:-|:-|:-|:-|:-|:-|
|convnext_small_in22k|`Resize((640,480))`|`aug_transforms(size=(288,224), min_scale=0.75)`|0.0178*|01:51|0.0187
|convnext_small_in22k|`Resize((640,480))`|`aug_transforms(size=(320,240), min_scale=0.75)`|0.0202|02:07|0.0226
|convnext_small_in22k|`Resize(480, method='squish')`|`aug_transforms(size=224, min_scale=0.75)`|0.0211|01:30|0.0211
|convnext_small_in22k|`Resize(480)`|`aug_transforms(size=224, min_scale=0.75)`|0.0216|01:29|0.0202
|vit_small_patch16_224|`Resize(480)`|`aug_transforms(size=224, min_scale=0.75)`|0.0202*|00:44|0.0250
|vit_small_patch16_224|`Resize(480, method='squish')`|`aug_transforms(size=224, min_scale=0.75)`|0.0216|00:47|0.0245
|vit_small_patch16_224|`Resize(640, method=ResizeMethod.Pad, pad_mode=PadMode.Zeros)`|`aug_transforms(size=224, min_scale=0.75)`|0.0226|00:50|0.0221
|swinv2_base_window12_192_22k|`Resize(480, method='squish')`|`aug_transforms(size=192, min_scale=0.75)`|0.0163*|02:30|0.0173
|swinv2_base_window12_192_22k|`Resize(640, method=ResizeMethod.Pad, pad_mode=PadMode.Zeros)`|`aug_transforms(size=192, min_scale=0.75)`|0.0187|03:27|0.0192
|swinv2_base_window12_192_22k|`Resize(480)`|`aug_transforms(size=192, min_scale=0.75)`|0.0197|03:22|0.0183
|swin_small_patch4_window7_224|`Resize(480, method='squish')`|`aug_transforms(size=224, min_scale=0.75)`|0.0202*|01:48|0.0207
|swin_small_patch4_window7_224|`Resize(480)`|`aug_transforms(size=224, min_scale=0.75)`|0.0207|01:47|0.0231
|swin_small_patch4_window7_224|`Resize(640, method=ResizeMethod.Pad, pad_mode=PadMode.Zeros)`|`aug_transforms(size=224, min_scale=0.75)`|0.0221|01:54|0.0183

\* = lowest error rate for the architecture

### Preparing an Ensemble for Kaggle Submission

I'll retrain and create an ensemble of the top 3 models based on TTA Error Rate (First Run):

- swinv2_base_window12_192_22k (0.0163)
- convnext_small_in22k (0.0178)
- vit_small_patch16_224 (0.0202)

The swin_small_patch4_window7_224 models did not outperform the quicker/smaller vit model so I won't use them in this submission.

Later on in the video, Jeremy walks through an example of how he trained large versions of the small models he tested. In this section, he used the following training function, which I'll use here for these small models, to prepare my submission predictions. Note that Jeremy has removed `seed=42` since in the ensemble for submission we want to use different validation sets when training each model (whereas before we wanted to use the same validation set to better compare the performance between models). I've also changed a couple of things (I'm not exporting the models, and I'm using a smaller batch size).

In [15]:
# store the tta predictions in a list
tta_res = []

In [16]:
# run this once and re-use for all trainings
tst_files = get_image_files(path/'test_images')
tst_files.sort()

In [17]:
def train(arch, item, batch, accum=False):
    kwargs = {'bs': 16} if accum else {}
    dls = ImageDataLoaders.from_folder(trn_path, valid_pct=0.2, item_tfms=item, batch_tfms=batch, **kwargs)
    cbs = GradientAccumulation(2) if accum else []
    learn = vision_learner(dls, arch, metrics=error_rate, cbs=cbs).to_fp16()
    learn.fine_tune(12, 0.01)
    
    # TTA predictions using test dataset
    tst_dl = dls.test_dl(tst_files)
    tta_res.append(learn.tta(dl=dls.test_dl(tst_files)))

In [18]:
arch = 'swinv2_base_window12_192_22k'

In [19]:
train(arch, item=Resize(480, method='squish'), batch=aug_transforms(size=192, min_scale=0.75))

In [20]:
len(tta_res), len(tta_res[0][0])

(1, 3469)

In [21]:
arch = 'convnext_small_in22k'

In [22]:
train(arch, item=Resize((640,480)), batch=aug_transforms(size=(288,224), min_scale=0.75))

Downloading: "https://dl.fbaipublicfiles.com/convnext/convnext_small_22k_224.pth" to /root/.cache/torch/hub/checkpoints/convnext_small_22k_224.pth


In [23]:
len(tta_res), len(tta_res[0][0]), len(tta_res[1][0])

(2, 3469, 3469)

In [24]:
arch = 'vit_small_patch16_224'

In [25]:
train(arch, item=Resize(480), batch=aug_transforms(size=224, min_scale=0.75))

In [26]:
len(tta_res), len(tta_res[0][0]), len(tta_res[1][0]), len(tta_res[2][0])

(3, 3469, 3469, 3469)

Before I stack the predictions and prepare them for the submission, I'll save the list of predictions:

In [ ]:
save_pickle('/kaggle/working/tta_res.pkl', tta_res)

Next, I'll take a quick detour and follow the steps Jeremy shares in [Live Coding 11](https://youtu.be/j-zMF2VirA8?feature=shared&t=3311).

First, he takes the first item from each list in `tta_res` (the predictions) and stores them in a list called `tta_prs`. The list returned by `learn.tta` has a second item of `None`, which represents the targets (which we don't have in the test set) we need to pick out just the first item (the predictions).

`zip`ping the items in `tta_res` creates a list of two tuples: a tuple with the three sets of predictions (the first item in each element of `tta_res`) and a tuple with three `None`s (the second item of each element of `tta_res`).

Here's a toy example to illustrate:

In [27]:
list(zip(*[[(1), None],[(2), None]]))

[(1, 2), (None, None)]

The `first` function rerurns the first element of an iterable object.

In [40]:
first??

Signature: first(x, f=None, negate=False, **kwargs)
Source:   
def first(x, f=None, negate=False, **kwargs):
    "First element of `x`, optionally filtered by `f`, or None if missing"
    x = iter(x)
    if f: x = filter_ex(x, f=f, negate=negate, gen=True, **kwargs)
    return next(x, None)
File:      /opt/conda/lib/python3.10/site-packages/fastcore/basics.py
Type:      function

In [28]:
first(list(zip(*[[(1), None],[(2), None]])))

(1, 2)

The second element of the zipped `tta_res` list is a tuple of `None`s.

In [29]:
list(zip(*tta_res))[1]

(None, None, None)

I'll now apply this code to `tta_res`:

In [30]:
tta_prs = first(zip(*tta_res))

In [31]:
len(tta_prs[0])

3469

Next, in order to take the mean value of the predictions, we `stack` them into a tensor:

In [32]:
t_tta = torch.stack(tta_prs)

In [33]:
t_tta.shape

torch.Size([3, 3469, 10])

Then, we take the mean of the three predictions for each of the 10 classes for each image.

In [34]:
avg_pr = t_tta.mean(0)

In [35]:
avg_pr.shape

torch.Size([3469, 10])

We then get the index of the largest probability out of the 10 classes for each image, which is the "prediction" that the model has made for the image.

In [36]:
idxs = avg_pr.argmax(dim=1)
idxs.shape

torch.Size([3469])

In [37]:
idxs

tensor([7, 8, 3,  ..., 8, 1, 5])

Finally, we convert those indexes to strings of disease names using the `vocab` and prepare the submission file:

In [38]:
dls = ImageDataLoaders.from_folder(trn_path, valid_pct=0.2, item_tfms=Resize(224))
mapping = dict(enumerate(dls.vocab))
ss = pd.read_csv(path/'sample_submission.csv')
results = pd.Series(idxs.numpy(), name='idxs').map(mapping)
ss.label = results
ss.to_csv('ensemble_subm.csv', index=False)

In [39]:
!head ensemble_subm.csv

image_id,label
200001.jpg,hispa
200002.jpg,normal
200003.jpg,blast
200004.jpg,blast
200005.jpg,blast
200006.jpg,brown_spot
200007.jpg,dead_heart
200008.jpg,brown_spot
200009.jpg,hispa


## Gradient accumulation to prevent out of memory

If you run out of memory while training any of these large models, you can use `GradientAccumulation` to lower the memory usage. In the training loop we get the gradients, we add the gradients times the learning rate to the weights, and then we zero the gradients. What you could do is halve the batch size, so for example from 64 to 32, and then only zero the gradients every two iterations, and only do the update every two iterations. So you calculate in two batches what you calculate in one batch and it will be mathematically identical, That's called `GradientAccumulation`, which added to the `Learner` as a callback, which are things that change the behavior of the training. 

How batches work: we randomly shuffle the dataset, and grab the next batch size of images, we resize them all to be the same size, and we stack them on top of each other. If it's black and white images for example, we would have 64 (or whatever the batch size is) 640 x 480 (or whatever image size you want) images so we end up with a 64 x 640 x 480 tensor. Pretty much all of the functionality provided by PyTorch will work fine for a mini batch of things just as it would for a single thing.

Inference is often done on CPU instead of GPU since you only need to process one thing at a time. Or people will queue a few of them up and stick them on a GPU. 

In my [next blog post](https://vishalbakshi.github.io/blog/posts/2024-02-05-paddy-part-4/) I walk through the discussion and code from Live Coding 11.